### PR TITLE
MPC tweaks

### PR DESCRIPTION
**Pull request by tombrazier**
### Description

Fixed a bug in MPC tuning when MPC_MAX is not 255.

Modified a comment to make it clear that `M306 T` is best done at around first layer height.

Added the ability to set filament heat capacity at runtime.

### Requirements

MPCTEMP enabled.

### Benefits

Mainly the ability to tune MPC for different filaments without rebuilding the code.

### Related Issues

Repeatedly requested on previous MPC PRs and on discord.

### CODE REVIEW

**Comment by rq3** [comment]
@tombrazier, I'm currently running the bugfix from 5/22/2022, with MPC and the default PLA filament heat capacity. So far today I have done prints in PLA, PETG, and polycarbonate (at 300C), all with the same default PLA heat capacity.

Maybe a note in configuration.h is in order, to the effect that the default filament heat capacity is probably fine, and changing it is for "expert users" who are printing with weird materials at high speeds? I can envision users obsessing over this unnecessarily, but applaud your attention to detail.

File: Marlin/src/gcode/temp/M306.cpp [hunk 1]
```diff
@@ -36,14 +36,15 @@
  *  C<joules/kelvin>          Block heat capacity.
  *  E<extruder>               Extruder number to set. (Default: E0)
  *  F<watts/kelvin>           Ambient heat transfer coefficient (fan on full).
+ *  M<joules/kelvin/mm>       Filament heat capacity per mm.
```
Comment (thinkyhead) [inline Marlin/src/gcode/temp/M306.cpp]: `M` may not be used as a parameter name.

**Comment by thinkyhead** [comment]
I changed the `M` parameter to `H` because it's standard practice to avoid `M` (also, `G`) as parameter names since these are used as command letters. This is a legacy issue, since the parser can handle an `M` or `G` parameter in any command. But, we don't want to mess up G-code standards too badly for firmwares that might be using a different kind of parser.

**Comment by tombrazier** [comment]
> I changed the M parameter to H because it's standard practice to avoid M (also, G) as parameter names

Makes sense. https://github.com/MarlinFirmware/MarlinDocumentation/pull/442 will also need to be changed.

**Comment by thisiskeithb** [comment]
> > I changed the M parameter to H because it's standard practice to avoid M (also, G) as parameter names
> 
> Makes sense. [MarlinFirmware/MarlinDocumentation#442](https://github.com/MarlinFirmware/MarlinDocumentation/pull/442) will also need to be changed.

@CRCinAU: See above.

**Comment by CRCinAU** [comment]
> > > I changed the M parameter to H because it's standard practice to avoid M (also, G) as parameter names
> > 
> > 
> > Makes sense. [MarlinFirmware/MarlinDocumentation#442](https://github.com/MarlinFirmware/MarlinDocumentation/pull/442) will also need to be changed.
> 
> @CRCinAU: See above.

Updated.

**Comment by tombrazier** [comment]
> [wrap MPC_MAX](https://github.com/MarlinFirmware/Marlin/pull/24253/commits/99dd99d497bdf5deea5af13843e93378fb5ea251)

Self duly kicked.

**Comment by tombrazier** [comment]
> Maybe a note in configuration.h is in order, to the effect that the default filament heat capacity is probably fine, and changing it is for "expert users" who are printing with weird materials at high speeds? I can envision users obsessing over this unnecessarily, but applaud your attention to detail.

But obsessing over immaterial details is a huge part of the fun of 3D printing! I think I might add the note you mention to https://marlinfw.org/docs/features/model_predictive_control.html as it will be a bit wordy for a comment in Configuration.h.

**Comment by sricha1217** [comment]
Not sure if this has been mentioned elsewhere: it would be interesting to take the pwm values generated while a hotend was being controlled using PID and feed them in to the model to estimate the temperature at the heater. 
I suspect that if one were to compare that to the estimate heater temperatures when in MPC mode, the improvement in stability would be even more dramatic than when comparing the temperatures measured all the way out at the sensor.

**Comment by tombrazier** [comment]
@sricha1217 I think I do not really understand what you are saying.

PID does not run concurrently with MPC - it is one or the other. When MPC is running, it does feed the PWM output back into the model to estimate hotend temperature. MPC only uses sensor data to keep its model in line with reality, which is necessary because no model is perfect. But the cleverness is that MPC's modeled sensor and the real sensor exhibit the same latency so the "all the way out at the sensor" effect is all but nullified.

Have you read [this](https://marlinfw.org/docs/features/model_predictive_control.html)?

**Comment by sricha1217** [comment]
Thanks for the response.  Yes, I’ve read the article and I understand how MPC works very well.  I tried several times to reword my earlier comment, but it always came out just as confusing.  My suggestion would only serve to illustrate that the stability improvement of MPC over PID is even more dramatic than the sensor data suggests. But I’m not actually proposing an improvement or highlight a bug.  So I wouldn’t be at all offended if my comments were removed as being too far off topic.

**Comment by tombrazier** [comment]
Oh, yes, I suspect you are right. I have been thinking about the Revo hotend which appears to have a relatively loose coupling between heat block and nozzle. So the sensor temperature can vary quite a bit from the nozzle temperature. MPC could, in principle, control for the nozzle temperature on the Revo rather than for the heat block temperature. That would result in the reported temperatures appearing to be off target as they are reported from the sensor. I guess it would be interesting to mod the M105 output to report the MPC modeled temperature and see what it looks like.

**Comment by sricha1217** [comment]
Thanks again for the response- and fantastic work!  I suspect that the tightened control on temperature will take some peoples prints to the next level of “cleanness” due to more consistent viscosity. (Mine are still bumpy but for many reasons other than temperature.)
